### PR TITLE
Flink: Project the RowData to remove meta-columns

### LINKED ISSUE #2730: Encounter exceptions when query the iceberg table filled with change logs by using flink SQL
Saying if we write few change log records ( INSERT,  DELETE, UPDATE_BEFORE, UPDATE_AFTER ) into apache iceberg table and then query it by using flink SQL, then we will encounter the exception: 

```txt
[flink-akka.actor.default-dispatcher-2] INFO org.apache.flink.runtime.executiongraph.ExecutionGraph - Source: Custom Source -> SinkConversionToTuple2 -> Sink: Select table sink (1/1) (03d66a7324d07ea31972ce3cc8d3f1df) switched from RUNNING to FAILED on 8e8e9f3c-67a5-4aaa-9cc8-18400ceee4a3 @ localhost (dataPort=-1).
java.lang.IllegalArgumentException: Row arity: 4, but serializer arity: 2
	at org.apache.flink.table.runtime.typeutils.RowDataSerializer.copy(RowDataSerializer.java:124)
	at org.apache.flink.table.runtime.typeutils.RowDataSerializer.copy(RowDataSerializer.java:48)
	at org.apache.flink.streaming.runtime.tasks.CopyingChainingOutput.pushToOperator(CopyingChainingOutput.java:69)
	at org.apache.flink.streaming.runtime.tasks.CopyingChainingOutput.collect(CopyingChainingOutput.java:46)
	at org.apache.flink.streaming.runtime.tasks.CopyingChainingOutput.collect(CopyingChainingOutput.java:26)
	at org.apache.flink.streaming.api.operators.CountingOutput.collect(CountingOutput.java:50)
	at org.apache.flink.streaming.api.operators.CountingOutput.collect(CountingOutput.java:28)
	at org.apache.flink.streaming.api.operators.StreamSourceContexts$ManualWatermarkContext.processAndCollect(StreamSourceContexts.java:317)
	at org.apache.flink.streaming.api.operators.StreamSourceContexts$WatermarkContext.collect(StreamSourceContexts.java:411)
	at org.apache.flink.streaming.api.functions.source.InputFormatSourceFunction.run(InputFormatSourceFunction.java:92)
	at org.apache.flink.streaming.api.operators.StreamSource.run(StreamSource.java:110)
	at org.apache.flink.streaming.api.operators.StreamSource.run(StreamSource.java:66)
	at org.apache.flink.streaming.runtime.tasks.SourceStreamTask$LegacySourceFunctionThread.run(SourceStreamTask.java:241)
```

The way to reproduce this bug is quite easy ,  please just apply the following patch, and then run the unit tests  `org.apache.iceberg.flink.TestChangeLogTable.testSqlChangeLogOnIdKey`

```patch
diff --git a/flink/src/test/java/org/apache/iceberg/flink/TestChangeLogTable.java b/flink/src/test/java/org/apache/iceberg/flink/TestChangeLogTable.java
index d44f45ab5..f4334b43c 100644
--- a/flink/src/test/java/org/apache/iceberg/flink/TestChangeLogTable.java
+++ b/flink/src/test/java/org/apache/iceberg/flink/TestChangeLogTable.java
@@ -273,6 +273,8 @@ public class TestChangeLogTable extends ChangeLogTableTestBase {
     Table table = createTable(tableName, key, partitioned);
     sql("INSERT INTO %s SELECT * FROM %s", tableName, SOURCE_TABLE);
 
+    sql("SELECT * FROM %s", tableName);
+
     table.refresh();
     List<Snapshot> snapshots = findValidSnapshots(table);
     int expectedSnapshotNum = expectedRecordsPerCheckpoint.size();
```

I guess other query engines such as spark sql , presto sql, hive sql will also encounter this issue if we've write few change log events in the iceberg table.

I remember there's another issue also encountered this exception ( I encountered it many times in our asia user group). Will prepare a  pull request for this.

### PR DESCRIPTION
**Pull request by Reo-LEI**
This PR is completed on the basis of #2731 and trying to fixes #2730. Thanks for the contribution of @openinx.

In this PR, I make `RowDataProjection` as row data wrapper as this comment https://github.com/apache/iceberg/pull/2731#discussion_r669313152 mentioned and supprot the `Map` and `List` type projection.

### CODE REVIEW

**Comment by Reo-LEI** [comment]
@rdblue @kbendick @openinx @stevenzwu Could you take a look of this again? 😄

**Comment by openinx** [comment]
Let me take another look today !  Thanks @Reo-LEI for the updating.

File: flink/src/main/java/org/apache/iceberg/flink/data/RowDataProjection.java [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+import org.apache.iceberg.flink.FlinkSchemaUtil;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.types.Types;
+
+public class RowDataProjection implements RowData {
+  /**
+   * Creates a projecting wrapper for {@link RowData} rows.
+   * <p>
+   * This projection does not work with repeated types like lists and maps.
```
Comment (openinx) [inline flink/src/main/java/org/apache/iceberg/flink/data/RowDataProjection.java:42]: This projection does not work with repeated types like lists and maps with nested children types ? I think it works fine for lists/maps with primitive children types.

File: flink/src/main/java/org/apache/iceberg/flink/data/RowDataProjection.java [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+   * @return a wrapper to project rows
+   */
+  public static RowDataProjection create(Schema schema, Schema projectedSchema) {
+    return RowDataProjection.create(FlinkSchemaUtil.convert(schema), schema.asStruct(), projectedSchema.asStruct());
+  }
+
+  /**
+   * Creates a projecting wrapper for {@link RowData} rows.
+   * <p>
+   * This projection does not work with repeated types like lists and maps.
```
Comment (openinx) [inline flink/src/main/java/org/apache/iceberg/flink/data/RowDataProjection.java:55]: ditto.

File: flink/src/main/java/org/apache/iceberg/flink/data/RowDataProjection.java [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+    }
+
+    this.getters = new RowData.FieldGetter[projectType.fields().size()];
+    for (int i = 0; i < getters.length; i++) {
+      Types.NestedField projectField = projectType.fields().get(i);
+      Types.NestedField rowField = rowStruct.field(projectField.fieldId());
+      if (rowField == null) {
+        throw new IllegalArgumentException(String.format(
+            "Cannot locate the project field <%s> in the iceberg struct <%s>", projectField, rowStruct));
+      }
```
Comment (openinx) [inline flink/src/main/java/org/apache/iceberg/flink/data/RowDataProjection.java:82]: Nit: 

```java
      Preconditions.checkNotNull(rowField,
          "Cannot locate the project field <%s> in the iceberg struct <%s>", projectField, rowStruct);
```

File: flink/src/main/java/org/apache/iceberg/flink/data/RowDataProjection.java (hunk 1 above)
Comment (openinx) [reply flink/src/main/java/org/apache/iceberg/flink/data/RowDataProjection.java:42]: It's better to say: `Projecting a partial map key or value with non-primitive type does not work in this projection wrapper`

File: flink/src/main/java/org/apache/iceberg/flink/source/RowDataFileScanTaskReader.java [hunk 4]
```diff
@@ -70,9 +71,17 @@ public RowDataFileScanTaskReader(Schema tableSchema, Schema projectedSchema,
...
     FlinkDeleteFilter deletes = new FlinkDeleteFilter(task, tableSchema, projectedSchema, inputFilesDecryptor);
-    return deletes
-        .filter(newIterable(task, deletes.requiredSchema(), idToConstant, inputFilesDecryptor))
-        .iterator();
+    CloseableIterable<RowData> iterable = deletes.filter(
+        newIterable(task, deletes.requiredSchema(), idToConstant, inputFilesDecryptor)
+    );
+
+    // Project the RowData to remove the extra meta columns.
+    if (!projectedSchema.sameSchema(deletes.requiredSchema())) {
```
[the remark this answers is not in the document]
Comment (openinx) [reply flink/src/main/java/org/apache/iceberg/flink/source/RowDataFileScanTaskReader.java:80]: The correct approach to address the above cases in unit tests is :  we create a separate unit case with an iceberg table, the iceberg table can be v1 table or v2 table. For the v1 table, we just write few records into it. Finally we use the `FlinkInputFormat` or `DataIterator`  to open this table and assert the returned `RowData`s are not a `RowDataProjection` type.  For the v2 table,  the difference is the returned `RowData` is a `RowDataProjection` type.

File: flink/src/test/java/org/apache/iceberg/flink/TestChangeLogTable.java [hunk 5]
```diff
@@ -261,7 +262,7 @@ private Table createTable(String tableName, List<String> key, boolean isPartitio
   private void testSqlChangeLog(String tableName,
                                 List<String> key,
                                 List<List<Row>> inputRowsPerCheckpoint,
-                                List<List<Record>> expectedRecordsPerCheckpoint) throws Exception {
+                                List<List<Row>> expectedRecordsPerCheckpoint) throws Exception {
```
[the remark this answers is not in the document]
Comment (openinx) [reply flink/src/test/java/org/apache/iceberg/flink/TestChangeLogTable.java:265]: Yeah, the returned data from flink SQL query are collection with `Row` types,   it's good to convert all the `Record` to `Row` for equality checking purpose.

File: flink/src/main/java/org/apache/iceberg/flink/data/RowDataProjection.java (hunk 1 above)
Comment (Reo-LEI) [reply flink/src/main/java/org/apache/iceberg/flink/data/RowDataProjection.java:42]: The meaning of this comment is exactly what you said that the projection will not project the nested children types of repeated types. I will rephrase it.

File: flink/src/test/java/org/apache/iceberg/flink/data/TestRowDataProjection.java [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+
+  private void generateAndValidate(Schema schema, Schema projectSchema) {
+    int numRecords = 100;
+    Iterable<Record> recordList = RandomGenericData.generate(schema, numRecords, 102L);
+    Iterable<RowData> rowDataList = RandomRowData.generate(schema, numRecords, 102L);
+
+    StructProjection structProjection = StructProjection.create(schema, projectSchema);
+    RowDataProjection rowDataProjection = RowDataProjection.create(schema, projectSchema);
+
+    Iterator<Record> recordIter = recordList.iterator();
```
[the remark this answers is not in the document]
Comment (openinx) [reply flink/src/test/java/org/apache/iceberg/flink/data/TestRowDataProjection.java:316]: The issue is: currently the `RandomRowData.generate` only returns `Iterator<RowData>`, it does not return the `List<RowData>`.  So we have to use the `Iterable` & `Iterator` here.

File: flink/src/main/java/org/apache/iceberg/flink/source/RowDataFileScanTaskReader.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (openinx) [reply flink/src/main/java/org/apache/iceberg/flink/source/RowDataFileScanTaskReader.java:80]: Let's make it to be a separate issue: https://github.com/apache/iceberg/issues/3431

**Review by openinx — APPROVED** [review]
Looks good to me overall, thanks @Reo-LEI for the great contribution and thanks @stevenzwu for the double check.  I left several minor comments.

**Comment by Reo-LEI** [comment]
I adressed the rest of comments just now, and you can check this again @openinx. And thanks @openinx @stevenzwu @kbendick @rdblue for review,

File: api/src/main/java/org/apache/iceberg/util/StructProjection.java [hunk 7]
```diff
@@ -155,6 +155,10 @@ public int size() {
 
   @Override
   public <T> T get(int pos, Class<T> javaClass) {
+    if (struct == null) {
+      return null;
```
[the remark this answers is not in the document]
Comment (rdblue) [reply api/src/main/java/org/apache/iceberg/util/StructProjection.java:161]: I agree with the behavior here. Thanks for pointing to #2738.

In the future, I think it is better to fix bugs like this in separate PRs that address a specific issue, like #2738. Having this mixed into a different PR means we can't cleanly cherry-pick it and we lose information about when the other issue was fixed. This is minor, but I think it really helps for maintainability in the long run.

**Comment by kbendick** [comment]
Sorry for missing some pings. Was out of office for a few weeks a while back and have still been playing a bit of catch up.

Please feel free to message me on slack if it's urgent btw. But retroactive +1. 🙂

File: api/src/main/java/org/apache/iceberg/util/StructProjection.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (openinx) [reply api/src/main/java/org/apache/iceberg/util/StructProjection.java:161]: Agree, It is very necessary to implement sufficient focused feature within an independent PR.

[the remark this answers is not in the document]
Comment (Reo-LEI) [reply api/src/main/java/org/apache/iceberg/util/StructProjection.java:161]: OK, I will separate it in next time. Thanks for bring this up!

**Pull request merged**